packages/hebao_v1/contracts/modules/security/GuardianUtils.sol

```diff
@@ -0,0 +1,176 @@
+            require(!walletOwnerSigned, "WALLET_OWNER_SIGNATURE_NOT_ALLOWED");
+        }
+
+        // If no social guardians are used do self authentication
+        if (numFriends + numFamily + numSocialOther == 0) {
+            /* Self authentication */
+            uint numSelfControlledRequired = 2;
+            // If the owner is allowed to sign:
+            // - increase the total number of self controlled guardians
+            // - increase the number of required self controlled guardians
+            if (requirement != SigRequirement.OwnerNotAllowed) {
+                totalNumSelfControlled += 1;
+                numSelfControlledRequired += 1;
+            }
+            // - Either the number of required signers is reached
+            // - or the required number of signers is lower than the total number of self
+            //   controlled guardians, in which case all self controlled guardians need to sign
+            //   (which may include the wallet owner if he is allowed to sign).
+            require(
+                numSelfControlled >= numSelfControlledRequired ||
+                numSelfControlled == totalNumSelfControlled && totalNumSelfControlled > 0,
+                "NOT_ENOUGH_SIGNERS"
+            );
+        } else {
+            /* Social authentication */
+            uint numGroups = 0;
+            uint numGroupsWithMajority = 0;
+            // Count the number of active groups and see which ones have a majority of signers
+            if (totalNumFriends > 0) {
+                numGroups++;
+                numGroupsWithMajority += hasMajority(numFriends, totalNumFriends) ? 1 : 0;
+            }
+            if (totalNumFamily > 0) {
+                numGroups++;
+                numGroupsWithMajority += hasMajority(numFamily, totalNumFamily) ? 1 : 0;
+            }
+            if (totalNumSocialOther > 0) {
+                numGroups++;
+                numGroupsWithMajority += hasMajority(numSocialOther, totalNumSocialOther) ? 1 : 0;
+            }
+            // Require a majority of groups to have a majority of signing guardians
+            require(hasMajority(numGroupsWithMajority, numGroups), "NOT_ENOUGH_SIGNERS");
+        }
+    }
+
+    function hasMajority(
+        uint count,
+        uint total
+        )
+        internal
+        pure
+        returns (bool)
+    {
+        return (count >= (total / 2) + 1);
+    }
+
+    function countGuardians(
+        SecurityStore.Guardian[] memory guardians
+        )
+        internal
+        pure
+        returns (
+            uint numSelfControlled,
+            uint numFriends,
+            uint numFamily,
+            uint numSocialOther
+        )
+    {
+        for (uint i = 0; i < guardians.length; i++) {
+            if (guardians[i].types & GuardianType.SelfControlled() != 0) {
+                numSelfControlled++;
+            } else {
+                if (guardians[i].types & GuardianType.Friend() != 0) {
```

Comment:
if we simply do `if (guardians[i].types == GuardianType.Friend())` can we use enum for types?